File: packages/ld-adapter/src/__tests__/rate-limited-client.test.ts
import { describe, expect, it, vi } from 'vitest';
import {
	ApprovalRequiredError,
	LdRateLimitError,
} from '../errors/ld-adapter-error.js';
import {
	computeRetryDelayMs,
	createRateLimitedLdClient,
} from '../client/rate-limited-client.js';
import type { LaunchDarklyRawClient } from '../client/ld-api-client.js';

const rawClient = {
	flagsApi: {},
	config: {
		accessToken: 'test',
		baseUrl: 'https://app.launchdarkly.com',
		apiVersion: '20240415',
	},
} as LaunchDarklyRawClient;

describe('createRateLimitedLdClient', () => {
	it('passes successful calls through without retry', async () => {
		const client = createRateLimitedLdClient(rawClient);
		const fn = vi.fn().mockResolvedValue('ok');
		await expect(client.schedule(fn)).resolves.toBe('ok');
		expect(fn).toHaveBeenCalledOnce();
	});

	it('retries 429 honoring Retry-After header', async () => {
		const sleep = vi.fn().mockResolvedValue(undefined);
		const client = createRateLimitedLdClient(rawClient, {
			sleep,
			retries: 2,
			jitterMs: 0,
		});
		const fn = vi
			.fn()
			.mockRejectedValueOnce({
				status: 429,
				response: { header: { 'retry-after': '1' } },
			})
			.mockResolvedValueOnce('ok');

		await expect(client.schedule(fn)).resolves.toBe('ok');
		expect(fn).toHaveBeenCalledTimes(2);
		expect(sleep).toHaveBeenCalled();
		expect(computeRetryDelayMs({ status: 429, response: { header: { 'retry-after': '1' } } }, 0)).toBe(1000);
	});

	it('does not retry 422 fail-fast errors', async () => {
		const client = createRateLimitedLdClient(rawClient);
		const fn = vi.fn().mockRejectedValue({ status: 422, message: 'invalid' });
		await expect(client.schedule(fn)).rejects.toMatchObject({ status: 422 });
		expect(fn).toHaveBeenCalledOnce();
	});

	it('maps 405 to ApprovalRequiredError', async () => {
		const client = createRateLimitedLdClient(rawClient);
		const fn = vi.fn().mockRejectedValue({ status: 405 });
		await expect(client.schedule(fn)).rejects.toBeInstanceOf(
			ApprovalRequiredError,
		);
	});

	it('throws LdRateLimitError after retries exhausted on 429', async () => {
		const sleep = vi.fn().mockResolvedValue(undefined);
		const client = createRateLimitedLdClient(rawClient, {
			sleep,
			retries: 1,
			jitterMs: 0,
		});
		const fn = vi.fn().mockRejectedValue({
			status: 429,
			response: { header: { 'retry-after': '1' } },
		});
		await expect(client.schedule(fn)).rejects.toBeInstanceOf(LdRateLimitError);
	});
});

File: kotlin/modules/ld-adapter/src/test/kotlin/com/ffpromo/ldadapter/client/RateLimitedClientTest.kt
package com.ffpromo.ldadapter.client

import com.ffpromo.ldadapter.errors.HttpResponseError
import com.ffpromo.ldadapter.errors.ApprovalRequiredError
import com.ffpromo.ldadapter.errors.LdRateLimitError
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import kotlin.random.Random

class RateLimitedClientTest {
    private val rawClient = LaunchDarklyRawClient(
        flagsApi = com.launchdarkly.api.api.FeatureFlagsApi(com.launchdarkly.api.ApiClient()),
        config = ResolvedLaunchDarklyConfig(
            accessToken = "test",
            baseUrl = "https://app.launchdarkly.com",
            apiVersion = "20240415",
        ),
        httpClient = okhttp3.OkHttpClient(),
    )

    @Test
    fun `passes successful calls through without retry`() = runTest {
        val client = createRateLimitedLdClient(rawClient)
        var calls = 0
        val result = client.schedule {
            calls++
            "ok"
        }
        assertEquals("ok", result)
        assertEquals(1, calls)
    }

    @Test
    fun `retries 429 honoring Retry-After header`() = runTest {
        var sleepCalls = 0
        val client = createRateLimitedLdClient(
            rawClient,
            RateLimitedLdClientOptions(
                sleep = { sleepCalls++ },
                retries = 2,
                jitterMs = 0,
                random = Random(0),
            ),
        )
        var calls = 0
        val result = client.schedule {
            calls++
            if (calls == 1) {
                throw HttpResponseError(
                    message = "rate limited",
                    status = 429,
                    headers = mapOf("retry-after" to "1"),
                )
            }
            "ok"
        }

        assertEquals("ok", result)
        assertEquals(2, calls)
        assertEquals(1, sleepCalls)
        assertEquals(
            1000L,
            computeRetryDelayMs(
                HttpResponseError("rate limited", 429, headers = mapOf("retry-after" to "1")),
                jitterMs = 0,
                random = Random(0),
            ),
        )
    }

    @Test
    fun `does not retry 422 fail-fast errors`() = runTest {
        val client = createRateLimitedLdClient(rawClient)
        var calls = 0
        val error = assertThrows<LdApiErrorWrapper> {
            client.schedule {
                calls++
                throw LdApiErrorWrapper("invalid", 422)
            }
        }
        assertEquals(422, error.status)
        assertEquals(1, calls)
    }

    @Test
    fun `maps 405 to ApprovalRequiredError`() = runTest {
        val client = createRateLimitedLdClient(rawClient)
        assertThrows<ApprovalRequiredError> {
            client.schedule {
                throw LdApiErrorWrapper("approval required", 405)
            }
        }
    }

    @Test
    fun `throws LdRateLimitError after retries exhausted on 429`() = runTest {
        val client = createRateLimitedLdClient(
            rawClient,
            RateLimitedLdClientOptions(
                sleep = {},
                retries = 1,
                jitterMs = 0,
                random = Random(0),
            ),
        )
        assertThrows<LdRateLimitError> {
            client.schedule {
                throw HttpResponseError(
                    message = "rate limited",
                    status = 429,
                    headers = mapOf("retry-after" to "1"),
                )
            }
        }
    }
}
